### PR TITLE
Filtering labels using tokenizer

### PR DESCRIPTION
**Pull request by yanenkoa**
@dbarashev Please review.

### CODE REVIEW

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 1]
```diff
@@ -162,9 +164,55 @@ define((require, exports, module) ->
     ch = ContextHandler = {
       contextPreviewExists: false
       UPDATE_DELAY: 1000
+      LABEL_SEQUENCE: [
+        {type: "keyword", value: /^\\label$/}
+        {type: "lparen", value: /^\{$/}
+        {type: "variable.parameter", value: /.*/}
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: is it true that all text inside label is variable.parameter even if it is actually another macro with its own name, brackets and arguments?

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 2]
```diff
@@ -162,9 +164,55 @@ define((require, exports, module) ->
...
+        tokenIterator.stepForward() # at first it should be on the closing token of begin sequence
+        {row: curRow, column: curColumn} = tokenIterator.getCurrentTokenPosition()
+        token = tokenIterator.getCurrentToken()
+
+        while range.contains(curRow, curColumn + token.value.length)
+          tokenToMatch = ch.LABEL_SEQUENCE[labelSequenceIndex]
+
+          if token.type == tokenToMatch.type and tokenToMatch.value.test(token.value)
+
+            if labelSequenceIndex == ch.LABEL_PARAMETER_INDEX
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: What do you think of more simple and probably more useful function which searches for a given keyword and removes that keyword and everything inside a properly balanced pair of curly brackets afterwards? I mean, given text `"\foo\bar\label{\foo{}\bar{}}\baz\qux"`, iterator positioned at the first token, and keyword `\label` it searches for `token.type=keyword` and `value=\label`, then runs `findSurroundingBrackets` from the position just after `\label{` and removes both `\label` and the range returned by `findSurroundingBrackets`

We coud reuse such code in other similar scenarios.

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 3]
```diff
@@ -162,9 +164,55 @@ define((require, exports, module) ->
...
+        labelParameters = []
+        curLabelParameter = null
+        curLabelTokens = []
+
+        tokenIterator = new TokenIterator(editor.getSession(), range.start.row, range.start.column)
+        tokenIterator.stepForward() # at first it should be on the closing token of begin sequence
+        {row: curRow, column: curColumn} = tokenIterator.getCurrentTokenPosition()
+        token = tokenIterator.getCurrentToken()
+
+        while range.contains(curRow, curColumn + token.value.length)
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: to restrict the scope you can pass a wrapper around TokenIterator which iterates until reaches the end of range

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 4]
```diff
@@ -162,9 +164,55 @@ define((require, exports, module) ->
...
-        editor.getSession().getTextRange(range)
+
+        tokenValues = []
+        labelSequenceIndex = 0
+        labelParameters = []
+        curLabelParameter = null
+        curLabelTokens = []
+
+        tokenIterator = new TokenIterator(editor.getSession(), range.start.row, range.start.column)
+        tokenIterator.stepForward() # at first it should be on the closing token of begin sequence
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: i think you can just pass a properly positioned TokenIterator as an argument and remove lines 183-184

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 1 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: It doesn't look like macros are even supported inside labels. It seems as though Katex, when parsing arguments, just searches for the first closing brace `}` and considers everything up to this as arguments. Which is actually different from main Papeeria compiler, where things like this:

    \label{abra{cadabra}}

are actually allowed and can be referenced with `\ref`. But even in main compiler, things like this

    \label{\begin{equation}\end{equation}}

result in an error. So, answering your question -- it **does** seem to always be a variable.parameter.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 4 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: It's not quite clear, how to determine a proper initial position for token iterator. I think whatever code I can write to determine initial position would just in some way duplicate code from TokenIterator class, and then what's the point.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 1 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Dunno anything about equations and environments inside labels, but simple macros definitely do work:
```
\documentclass{article}

\usepackage{amsthm}
\usepackage{hyperref}
\usepackage{cleveref}
  
\newtheorem{thm}[equation]{Theorem}
\newcommand{\foo}[1]{#1#1}
\begin{document}
\begin{thm}
\label{\foo{a}\foo{b}}
This is my theorem\'a
\end{thm}
As proven in theorem \ref{aabb}...
\end{document}
```

Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: I mean, it is perfectly okay from the TeX compiler perspective, but highlighter may think different and it does matter for us what it thinks because we depend on token types.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 4 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Sry, did not understand your suggestion fully. To answer your actual question, I don't really see a point in creating token iterator before calling this function, and then passing it here. In every case we call it, we just pass it `ch.curRange`, and wrap the start of it every time in token iterator doesn't really seem reasonable to me.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 1 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Huh. I'm going to need some time to figure it out.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 4 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: you don't need to determine it. You just want the caller to pass properly positioned iterator. You already implicitly request to pass quite specific range: one which starts after \begin{equation} -- this is not visible in the function contract but you rely on it and do `stepForward` here with fingers crossed. I suggest moving this responsibility on the caller: let those who call the function position the iterator and do `stepForward` if needed.

Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: I see. I understand the arguments, but that will make every call to this function really verbose. But yeah, I'll do it.

Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: I was mistaken, it is actually called only once.

**Comment by yanenkoa** [comment]
@dbarashev pushed some changes, please review.

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 5]
```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
...
+      @tokenIterator.stepBackward()
+      curToken = @tokenIterator.getCurrentToken()
+      {row: tokenRow, column: tokenColumn} = @tokenIterator.getCurrentTokenPosition()
+      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
+      if not @range.containsRange(tokenRange)
+        @tokenIterator.stepForward()
+        @expired = true
+        return null
+      else
+        return @tokenIterator.getCurrentToken()
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: `return curToken` ? Also, please swap `if` branches: `if @range.containsRange() then return curToken else...`

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 6]
```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
...
+      @tokenIterator.stepForward()
+      curToken = @tokenIterator.getCurrentToken()
+      {row: tokenRow, column: tokenColumn} = @tokenIterator.getCurrentTokenPosition()
+      tokenRange = new Range(tokenRow, tokenColumn, tokenRow, tokenColumn + curToken.value.length)
+      if not @range.containsRange(tokenRange)
+        @tokenIterator.stepBackward()
+        @expired = true
+        return null
+      else
+        return @tokenIterator.getCurrentToken()
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: same comment as above

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 7]
```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
...
+      { type: "rparen", value: "}" }
+    ]
+    @END_EQUATION_TOKEN_SEQUENCE: [
+      { type: "storage.type", value: "\\end" }
+      { type: "lparen", value: "{" }
+      { type: "variable.parameter", value: "equation" }
+      { type: "rparen", value: "}" }
+    ]
+
+    constructor: (@editor) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: add a comment: `# empty constructor` (to make it clear that it is not a forgotten piece of code)

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 8]
```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
...
+
+    equalTokens: (token1, token2) ->
+      if token1? and token2?
+        return token1.type == token2.type and token1.value == token2.value
+      else return if token1? or token2? then false else true
+
+    getEquationStart: (tokenIterator) ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: You need to handle the case when iterator goes out of range and fail gracefully. Maybe throw error from iterator, and catch it like this:
```
getEquationStart: (it) ->
  try
    return doGetEquationStart(it)
  catch e
    return null
```

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 9]
```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
...
+      # following cycle pushes tokenIterator to the end of
+      # beginning sequence, if it is inside one
+      for token in @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: You can use class name here: `EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE`

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 10]
```diff
@@ -2,104 +2,116 @@ define((require, exports, module) ->
...
+        else
+          curSequenceIndex = @constructor.BEGIN_EQUATION_TOKEN_SEQUENCE.length - 1
+          curEquationStart = null
+      return curEquationStart
+
+    getEquationEnd: (tokenIterator) ->
+      # following cycle pushes tokenIterator to the start of
+      # ending sequence, if it is inside one
+      for token in @constructor.END_EQUATION_TOKEN_SEQUENCE.slice(0).reverse()
+        if @equalTokens(token, tokenIterator.getCurrentToken())
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: any sense to leave `equalTokens` as standalone function? This will save us some CPU cycles and it may matter because it is called frequently

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 11]
```diff
@@ -172,45 +184,48 @@ define((require, exports, module) ->
       ]
       LABEL_PARAMETER_INDEX: 2
 
-      getWholeEquation: (range) ->
+      getMacrosArgumentRange: (session, argumentStartPos) ->
+        argumentRange = findSurroundingBrackets(session, argumentStartPos)
+        if argumentRange.mismatch
+          return null
+        else
+          return new Range(argumentRange.start.row, argumentRange.start.column + 1,
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: why +1 ?

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 12]
```diff
@@ -172,45 +184,48 @@ define((require, exports, module) ->
...
-
-          else
-            labelSequenceIndex = 0
-            for labelToken in curLabelTokens
-              tokenValues.push(labelToken.value)
-            curLabelTokens = []
+        tokenPosition = tokenIterator.getCurrentTokenPosition()
+
+        while token?
+          if not token?
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: isn't it guaranteed by the previous `while`?

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 13]
```diff
@@ -172,45 +184,48 @@ define((require, exports, module) ->
...
+            break
+
+          noPush = false
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: pls rename to `acceptToken = true`. Double negation `if not noPush` blows the mind

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 14]
```diff
@@ -172,45 +184,48 @@ define((require, exports, module) ->
...
+          if token.type == "storage.type" and token.value == "\\label"
+            curLine = session.getLine(tokenPosition.row)
+            bracketPosition = tokenPosition.column + "\\label".length
+            if curLine[bracketPosition] == "{"
+              argumentRange = ch.getMacrosArgumentRange(session, {row: tokenPosition.row, column: bracketPosition + 1})
+              if argumentRange?
+                noPush = true
+                labelParameters.push(session.getTextRange(argumentRange))
+                tokenIterator = new ConstrainedTokenIterator(session, range, argumentRange.end.row, argumentRange.end.column + 1)
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: maybe create method `ConstrainedtTokenIterator.stepTo` ?

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 11 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: So that it would be **after** the bracket. Should I comment this?

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 12 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Sorry, left from debugging

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 14 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Got everything, will do.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 11 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: The confusing part is that end column is not decremented. Is it because `findSurroundingBrackets` returns closed interval [start, end] while Range takes semi-opened [start, end) ?

**Comment by dbarashev** [comment]
Pls write a few tests for ConstrainedTokenIterator and getWholeEquation

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 11 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Basically, yes.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 8 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: What range? This function actually determines the range for equation. It doesn't really know, in which range it should operate. Also, default TokenIterator actually doesn't throw errors, when it hits the end or start of a document, it just returns nulls for everything. And I don't use our own ConstrainedTokenIterator for the reason above.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 11 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: then add a comment explaining this conversion :)

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 8 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: ah, I see, it is a plain TokenIterator here. Maybe add a condition on `getCurrentToken() != null` to the `while` loop and return null? Just not to fail with Null pointer when you access `value`of the current token

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 11 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Ok.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 8 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Yeah, it actually happened quite a lot when I was debugging it :). Ok, I'll add it.

**Comment by yanenkoa** [comment]
@dbarashev did the things you suggested, except for tests for `getWholeEquation`, they're on the way.

**Comment by yanenkoa** [comment]
@dbarashev pushed some more, commit messages are self-explanatory.

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 15]
```diff
@@ -5,10 +5,189 @@ define((require, exports, module) ->
...
+      catch e
+        return ["Error!", e]
+
+    initPopover: => setTimeout((=>
+      popoverPosition = @getPopoverPosition(@getEquationEnd())
+      [title, rendered] = @getCurrentFormula()
+      @popoverHandler.show(@getFormulaElement(), title, rendered, popoverPosition)
+    ), 0)
+
+    getEquationEnd: ->
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: `getEquationEndRow` ?

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 16]
```diff
@@ -5,10 +5,189 @@ define((require, exports, module) ->
...
+
+    updatePopover: ->
+      if @contextPreviewExists
+        [title, rendered] = @getCurrentFormula()
+        @popoverHandler.setContent(@getFormulaElement(), title, rendered)
+
+    updateCallback: =>
+      if @lastChangeTime?
+        curTime = Date.now()
+        @currentDelayedUpdateId = setTimeout(@updateCallback, ContextHandler.UPDATE_DELAY - (Date.now() - @lastChangeTime))
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: use `curTime` ? Also, any sense to check that lastChangeTime happened less than UPDATE_DELAY ago? If it happened long ago then you can just continue here

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 17]
```diff
@@ -53,49 +244,60 @@ define((require, exports, module) ->
...
-    equalTokens: (token1, token2) ->
-      if token1? and token2?
-        return token1.type == token2.type and token1.value == token2.value
-      else return if token1? or token2? then false else true
+    @equalTokens: (token1, token2) ->
+      return token1.type == token2.type and token1.value == token2.value
 
     getEquationStart: (tokenIterator) ->
+      # if tokenIterator initialy is on the empty line, it's current token is null
+      if not tokenIterator.getCurrentToken()?
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: huh, really? I thought `getCurrentToken() == null` meant "iterator reached the end of the sequence", no? How do you distinguish "empty line" from "end of iteration" otherwise?

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 18]
```diff
@@ -2,118 +2,333 @@ define((require, exports, module) ->
...
+        @createContextPreview()
+    ), 0)
+
+
+  class ConstrainedTokenIterator
+    constructor: (@session, @range, row, column) ->
+      @tokenIterator = new TokenIterator(@session, row, column)
+      curToken = @tokenIterator.getCurrentToken()
+      if not curToken?
+        @expired = false
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Normally "expired" is used when talking about time. Consider "outOfRange" or "isOut" or "isStopped/Terminated"

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 17 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: I don't. If it's beyond the end (or start) of a document, then I just return null in a `while` loop.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 16 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: This actually never happens. `@lastChangeTime` is only set, if we press anything during `UPDATE_DELAY` milliseconds after the first keypress, and then in `@updateCallback`, which never happens in more than `UPDATE_DELAY` milliseconds after any keypress, we set `@lastChangeTime` back to null.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 17 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Sorry? What do I see here in the comment: `getCurrentToken() == null` when tokeniterator is on the empty line. Is it true that every time when it reaches empty line during iteration its current token becomes null?

Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Nope. It is only the case, when token iterator is **initially** on an empty line. If we have an empty line ahead of us during the iteration, it is just skipped.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 16 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: it does happen because the only guarantee which setTimeout provides is that code will be executed no early than the specified delay. It may execute later.

https://developer.mozilla.org/en-US/docs/Web/API/WindowTimers/setTimeout#Reasons_for_delays_longer_than_specified

Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Oh. Okay, I'll add the check then.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 17 above)
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: but it might as well mean that iterator is already positioned at the end of iteration, e.g. after getEquationEnd call. I believe that nothing goes wrong because when this iterator is at the end of iteration, it can be safely stepForward'ed arbitrary number of times, but still this code looks suspicious. Can you write a few tests which will call this method with iterator positioned at null (both when it is empty line and end of iteration) and tests equations with empty lines inside?

Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Ok, will do.

**Comment by yanenkoa** [comment]
@dbarashev added tests you were talking about. They actually revealed a couple more problems, so I fixed them.

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 19]
```diff
@@ -251,14 +252,19 @@ define((require, exports, module) ->
       return token1.type == token2.type and token1.value == token2.value
 
     getEquationStart: (tokenIterator) ->
-      # if tokenIterator initialy is on the empty line, it's current token is null
+      # if tokenIterator is initialy on the empty line, its current token is null
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: nit: initially (`l` should be doubled)

File: lib/ace/ext/papeeria/katex-previewer.coffee [hunk 20]
```diff
@@ -251,14 +252,19 @@ define((require, exports, module) ->
...
         tokenIterator.stepForward()
-      # following cycle pushes tokenIterator to the end of
-      # beginning sequence, if it is inside one
-      for token in EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCE
-        if EquationRangeHandler.equalTokens(token, tokenIterator.getCurrentToken())
-          tokenIterator.stepForward()
+        # if tokenIterator.getCurrentToken() is still null, then we're at the end of a file
+        if not tokenIterator.getCurrentToken()?
+          return null
+      else
```
Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: This looks like a bug because if `tokenIterator.getCurrenttoken` is null then you won't run this loop which pushes iterator forward. This is actually not a bug provided that we do not allow for empty lines in \begin{equation}, however it needs to be documented (like, "If token iterator is positioned at empty line then it is certainly not in the \begin{equation}") or just rewritten so that it didn't look like a bug (remove else, run the loop unconditionally)

**Comment by dbarashev** [comment]
Add a test for iterator positioned at empty line?

**Comment by yanenkoa** [comment]
@dbarashev the test for empty line is already there.

File: lib/ace/ext/papeeria/katex-previewer.coffee (hunk 20 above)
Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: If the loop is run unconditionally, then there are problems with `\begin{equation}` positioned at the very beginning of a file or `\end{equation}` positioned at the very end of a file. 

Say, we position token iterator initially after the end of a file. First, we're looking for an end. We see, that current token is null, and move back one step. Then it is in the document already, and inside `\end{equation}` token sequence. Then the iterator is pushed to its start by a for loop, and then equation ending is found without any problems. But we should return null.

Comment (dbarashev) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: I see. Pls add a comment describing these cases

Comment (yanenkoa) [inline lib/ace/ext/papeeria/katex-previewer.coffee]: Okay

**Comment by dbarashev** [comment]
Cool, LGTM
